### PR TITLE
Switch to workstation GC in case of constrained CPU resources

### PR DESCRIPTION
**Pull request by luhenry**
Right now, if the user sets the configuration so that the server GC is
used, the server GC will be loaded even in conditions where we know the
workstation GC would fare better. An example of such conditions is
constrained environment where there is only 1 or less CPU or with very
low memory.

This can be harmful if users deploy the same projects on different kind
of platforms: deploying to a 20+ cores server and to Azure Functions
will require largely different configurations for the runtime.

There are already multiple ways for the user to specify to use the
server GC or not:
 - setting `COMPlus_gcServer` as an environment variable
 - setting `gcServer` in the configuration file
 - setting `System.GC.Server` passed to `coreclr_initialize`

Fix https://github.com/dotnet/coreclr/issues/23949

### CODE REVIEW

**Comment by stephentoub** [comment]
Is there any case with 1 CPU where server GC would perform better?  I'm wondering if we're going to find ourselves needing a "COMPlus_GCServerNoReallyIWantGCServer" environment variable :smile:

**Comment by luhenry** [comment]
Per https://github.com/dotnet/coreclr/issues/23949#issuecomment-482745078, it was already the behavior in .NET Fx. @Maoni0 is there any cases where you would expect server GC to perform better with 1 CPU only?

File: src/gc/unix/gcenv.unix.cpp [hunk 1]
```diff
@@ -741,6 +741,12 @@ uint32_t GCToOSInterface::GetCurrentProcessCpuCount()
     return g_currentProcessCpuCount;
 }
 
+uint32_t GCToOSInterface::GetCurrentProcessCpuLimit()
+{
+    uint32_t cpuLimit, cpuCount = GCToOSInterface::GetCurrentProcessCpuCount();
```
Comment (janvorli) [inline src/gc/unix/gcenv.unix.cpp]: A nit - can you please structure the code the same way as in the gcenv.os.cpp? It seems easier to read than this condensed version.

**Comment by Maoni0** [comment]
I don't see why GC needs to know about this CPU limit at all...it serves the exact same purpose as current process cpu count, correct?

File: src/gc/env/gcenv.os.h [hunk 2]
```diff
@@ -388,6 +388,11 @@ class GCToOSInterface
     //  The number of processors
     static uint32_t GetCurrentProcessCpuCount();
```
Comment (VSadov) [inline src/gc/env/gcenv.os.h]: It seems like `GetCurrentProcessCpuCount` should be sufficient to choose workstation (single heap) vs server (a heap per core) GC.

**Comment by luhenry** [comment]
@Maoni0 @VSadov I am changing back to using `GetCurrentProcessCpuCount`. I initially did this change before https://github.com/dotnet/coreclr/commit/576ac07df17d52fcbf24e472b3b4aa1edb0f61f9 thus the need to have it here.

**Comment by Maoni0** [comment]
so you are switching to WKS GC on the GC side but the VM side remains to recognize it as SVR GC, ie, g_heap_type is still GC_HEAP_SVR. and there are things dependent on the type (IsServerGC). for example the write barrier should be using the WKS version but now it remains the SVR version. 

it would be easier to recognize this on the VM side and just update g_heap_type there.

**Pull request merged**